ddl/column_type_change_test.go

```diff
@@ -2129,3 +2129,17 @@ func (s *testColumnTypeChangeSuite) TestCastToTimeStampDecodeError(c *C) {
 	// Normal cast datetime to timestamp can succeed.
 	tk.MustQuery("select timestamp(cast('1000-11-11 12-3-1' as date));").Check(testkit.Rows("1000-11-11 00:00:00"))
 }
+
+// https://github.com/pingcap/tidb/issues/25285.
+func (s *testColumnTypeChangeSuite) TestCastFromZeroIntToTimeError(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test;")
+
+	tk.MustExec("drop table if exists t;")
+	tk.MustExec("create table t (a int);")
+	tk.MustExec("insert into t values (0);")
+	tk.MustGetErrCode("alter table t modify column a date;", mysql.ErrTruncatedWrongValue)
```

Comment:
How about the sql_mode is ""?
The result in MySQL 8.0 as follows:
```
mysql> alter table t modify column a date;
Query OK, 1 row affected, 1 warning (0.09 sec)
Records: 1  Duplicates: 0  Warnings: 1
```

Reply:
Good catch. I only checked the mode `NO_ZERO_IN_DATE` and `NO_ZERO_DATE` before, these two mode does not affect the column type changing to time-related types.

Maybe the key point is the strict mode.